src/julia_gc.c

```diff
@@ -0,0 +1,641 @@
+                treap_rot_right(treap);
+                treap = &R(*treap);
+            } else {
+                treap_rot_left(treap);
+                treap = &L(*treap);
+            }
+        }
+    }
+}
+
+static int treap_delete(treap_t **treap, void *addr)
+{
+    while (*treap != NULL) {
+        int c = cmp_ptr(addr, (*treap)->addr);
+        if (c == 0) {
+          treap_delete_node(treap);
+          return 1;
+        } else if (c < 0) {
+          treap = &L(*treap);
+        } else {
+          treap = &R(*treap);
+        }
+    }
+    return 0;
+}
+
+static uint64_t xorshift_rng_state = 1;
+
+static uint64_t xorshift_rng()
+{
+    uint64_t x = xorshift_rng_state;
+    x = x ^ (x >> 12);
+    x = x ^ (x << 25);
+    x = x ^ (x >> 27);
+    xorshift_rng_state = x;
+    return x * (uint64_t) 0x2545F4914F6CDD1DUL;
+}
+
+
+static treap_t *bigvals;
+
+void *alloc_bigval(size_t size) {
+  void *result = malloc(size);
+  memset(result, 0, size);
+  treap_t *node = alloc_treap();
+  node->addr = result;
+  node->prio = xorshift_rng();
+  treap_insert(&bigvals, node);
+  return result;
+}
+
+void free_bigval(void *p) {
+  if (p) {
+    treap_delete(&bigvals, p);
+    free(p);
+  }
+}
+
+static jl_datatype_t *datatype_mptr;
+static jl_datatype_t *datatype_bag;
+static void *GapStackBottom = NULL;
+static size_t GapStackAlign = sizeof(Int);
+static jl_ptls_t JuliaTLS = NULL;
+
+#ifndef NR_GLOBAL_BAGS
+#define NR_GLOBAL_BAGS 20000L
+#endif
+
+static Bag *GlobalAddr[NR_GLOBAL_BAGS];
+static const Char *GlobalCookie[NR_GLOBAL_BAGS];
+static Int GlobalCount;
+
+
+
+/****************************************************************************
+**
+*F  AllocateBagMemory( <type>, <size> )
+**
+**  Allocate memory for a new bag.
+**
+**  'AllocateBagMemory' is an auxiliary routine for the Boehm GC that
+**  allocates memory from the appropriate pool. 'gc_type' is -1 if all words
+**  in the bag can refer to other bags, 0 if the bag will not contain any
+**  references to other bags, and > 0 to indicate a specific memory layout
+**  descriptor.
+**/
+void *AllocateBagMemory(int type, UInt size)
+{
+  // HOOK: return `size` bytes memory of TNUM `type`.
+  void *result = (void *) jl_gc_alloc(JuliaTLS, size, datatype_bag);
+  return result;
+}
+
+void InitMarkFuncBags (
+    UInt                type,
+    TNumMarkFuncBags    mark_func )
+{
+  // HOOK: set mark function for type `type`.
+}
+
+void InitSweepFuncBags (
+    UInt                type,
+    TNumSweepFuncBags    mark_func )
+{
+  // HOOK: set sweep function for type `type`.
+  // This is intended for weak pointer objects.
+}
+
+static void TryMark(void *p) 
+{
+  jl_value_t *p2 = jl_pool_base_ptr(p);
+  if (!p2)
+    p2 = treap_find(bigvals, p);
+  if (p2) {
+    jl_gc_queue_root(p2);
+  }
+}
+
+void CHANGED_BAG(Bag bag) {
+  // This is the simple Julia write barrier; jl_gc_wb() is more accurate
+  // but requires a pointer to the originating back, too, so we go with
```

Comment:
"back" -> "bag"? Also, what does "originating" mean here? After looking at `jl_gc_wb` vs `jl_gc_wb_back`, my understanding is that the former expects a pointer to the bag that was modified (which is `bag` here), and another to the `child` bag to which a reference was inserted. So, shouldn't the comment say e.g. "... but requires a pointer to the child bag which was inserted into `bag`"?

To do that, we'd have to add a variant of `CHANGED_BAG` with two arguments, and start using this; this sounds like a bigger change. Probably shouldn't worry about this for now...